clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java

```diff
@@ -0,0 +1,328 @@
+import org.apache.kafka.clients.consumer.internals.events.BackgroundEvent;
+import org.apache.kafka.clients.consumer.internals.events.RebalanceListenerInvokedEvent;
+import org.apache.kafka.clients.consumer.internals.events.RebalanceStartedEvent;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.common.message.ConsumerGroupHeartbeatResponseData.Assignment;
+import org.apache.kafka.common.message.ConsumerGroupHeartbeatResponseData.TopicPartitions;
+import org.apache.kafka.clients.consumer.internals.Utils.TopicPartitionComparator;
+import org.apache.kafka.common.utils.LogContext;
+import org.slf4j.Logger;
+
+import java.time.Duration;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.concurrent.BlockingQueue;
+
+import static org.apache.kafka.clients.consumer.internals.RebalanceStep.ASSIGN;
+import static org.apache.kafka.clients.consumer.internals.RebalanceStep.LOSE;
+import static org.apache.kafka.clients.consumer.internals.RebalanceStep.REVOKE;
+
+/**
+ * {@code AssignmentReconciler} performs the work of reconciling this consumer's partition assignment as directed
+ * by the consumer group coordinator. When the coordinator determines that a change to the partition ownership of
+ * the group is required, it will communicate with each consumer to relay its respective <em>target</em>
+ * assignment, that is, the set of partitions for which that consumer should now assume ownership. It is the then the
+ * responsibility of the consumer to work toward that target by performing the necessary internal modifications to
+ * satisfy the assignment from the coordinator. In practical terms, this means that it must first determine the set
+ * difference between the <em>{@link SubscriptionState#assignedPartitions() current assignment}</em> and the
+ * <em>{@link Assignment#assignedTopicPartitions() target assignment}</em>.
+ *
+ * <p/>
+ *
+ * Internally, reconciliation requires the following steps:
+ *
+ * <ol>
+ *     <li>
+ *         Calculate the partitions to revoke; these are the partitions in the current assignment that are
+ *         <em>not in</em> the target assignment
+ *     </li>
+ *     <li>
+ *         Send a {@link RebalanceListenerInvokedEvent} so that the application thread will execute the
+ *         {@link ConsumerRebalanceListener#onPartitionsRevoked(Collection)} callback method
+ *     </li>
+ *     <li>
+ *         On confirmation of the callback method execution, remove the revoked partitions from the
+ *         {@link SubscriptionState#assignFromSubscribed(Collection) current assignment}
+ *     </li>
+ *     <li>
+ *         Signal to the heartbeat request manager to perform a heartbeat acknowledgement with the group coordinator
```

Comment:
I don't fully understand this. My mental model for the client side implementation is the following. The client receives a new target assignment. It computes the difference between it and its current assignment. This gives the set of partitions to revoke and the partitions to assign. Then, it does the revocation followed by the assignment of those partitions. Finally, it acknowledges the target assignment.

Reply:
@philipnee is your question: Should the heartbeat request manager send a heartbeat after revocation, or should it wait to send it after assignment?

Reply:
My mental model was similar to yours @dajac , just the clarification that I was expecting that the client would acknowledge revocations with a HB before processing any new assignment (in the case of a client that receives partitions to revoke and new partitions in the same HB response) 

Ex. client that is part of a group and owns partition A, from the client point of view:
- receives HB response with assignment [A,B]
- notices new partition B 
   - client performs assignment (callbacks)
   - sends ack as HB with owned topicPartitions [A,B]
- receives HB response with assignment [B,C]
- notices revocation of A and new assignment C. 
   - Client does revocation (callback), **sends ack as HB with owned topicPartitions [B]** to ack the revocation asap
   - Client does assignment of C (callback), sends ack as HB with owned TopicPartitions [B,C]
   
This is my understanding of the expected flow, and the explanation for @kirktrue 's comment. Is it correct?

Reply:
Got it. I was mistaken then, but it's really a simplification on the client side so sounds good to me.

Reply:
Just to recap, it's fair to say then that the reconciliation process when there are revocations and new assignments involves a single interaction between client-server (HB req + process revocation/assignment + HB resp), vs. the double rebalance that takes places with the old protocol in the same scenario, correct?

Reply:
The code has been changed to perform callbacks for both revoke and assign in one shot, with no heartbeats between them, as discussed above.